File: lib/price.ts
/**
 * ราคาทอง
 *
 * ทำไมต้องปรับฐาน:
 *   ทองล่วงหน้า COMEX (GC=F) แพงกว่า spot ราว $35 — ถ้าเอามาคำนวณแนวรับแนวต้านตรง ๆ
 *   เส้นจะเพี้ยนทั้งกระดานเมื่อวาดลงกราฟ XAUUSD ของโบรกเกอร์
 *
 * จึงใช้ PAXG (โทเคนทองคำ 1 ออนซ์ ห่าง spot ~$5) เป็นแท่งเทียน
 *   + ปรับฐานด้วยราคา spot จริง -> เหลือคลาดเคลื่อน ~$1-2
 *   + brokerOffset -> ตรงกับกราฟที่ผู้ใช้ดูจริง
 *
 * ส่วนการวัดปฏิกิริยาข่าวและโปรไฟล์รายชั่วโมงใช้ GC=F เพราะเป็นตลาดทองจริง
 * ที่ข่าววิ่งเข้าโดยตรง และวัดแค่ "ส่วนต่าง" ระดับราคาจึงไม่สำคัญ
 */

import { isMarketOpen } from "./time";
import type { Bar } from "./types";

const SPOT_URL = "https://api.gold-api.com/price/XAU";
const PAXG_URL = "https://api.binance.com/api/v3/klines?symbol=PAXGUSDT&interval=15m&limit=1000";
const GCF_URL = "https://query1.finance.yahoo.com/v8/finance/chart/GC=F";

const UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64)";

/**
 * อายุ cache ของแต่ละแหล่ง — ตั้งให้พอดีกับความถี่ที่ข้อมูลเปลี่ยนจริง
 * ราคาปัจจุบันสดหน่อย ส่วนแท่งเทียนย้อนหลังไม่ต้อง เพราะแท่งใหม่เกิดทุก 15 นาทีอยู่แล้ว
 */
const TTL = { spot: 60, bars: 120, history: 300 };

async function getJson<T>(url: string, timeoutMs = 25_000, revalidate = TTL.bars): Promise<T> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(url, {
      headers: { "User-Agent": UA },
      signal: controller.signal,
      next: { revalidate },
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    return (await res.json()) as T;
  } finally {
    clearTimeout(timer);
  }
}

/** ราคา spot ทองคำล่าสุด (ดอลลาร์ต่อออนซ์) */
export async function fetchSpot(): Promise<number> {
  const data = await getJson<{ price: number }>(SPOT_URL, 25_000, TTL.spot);
  return Number(data.price);
}

/** แท่งเทียน 15 นาทีของ PAXG — ยังไม่ปรับฐาน */
export async function fetchPaxg15m(): Promise<Bar[]> {
  const rows = await getJson<unknown[][]>(PAXG_URL, 25_000, TTL.bars);
  return rows.map((row) => ({
    ts: Math.floor(Number(row[0]) / 1000),
    o: Number(row[1]),
    h: Number(row[2]),
    l: Number(row[3]),
    c: Number(row[4]),
  }));
}

interface YahooChart {
  chart: {
    result: {
      timestamp: number[];
      indicators: { quote: { open: (number | null)[]; high: (number | null)[]; low: (number | null)[]; close: (number | null)[] }[] };
      meta?: { regularMarketPrice?: number };
    }[];
  };
}

/**
 * แท่งเทียนทองล่วงหน้า COMEX
 * ขอบเขตที่ Yahoo ให้ (ทดสอบแล้ว): 1m=7วัน · 5m=60วัน · 15m=60วัน · 1h=2ปี
 */
/**
 * ตัดแท่งที่ timestamp ซ้ำ เก็บตัวที่มาทีหลัง
 *
 * Yahoo ส่งแท่งที่ยังไม่ปิดมาซ้ำ timestamp เดียวกับแท่งล่าสุด (เจอจริง: 14:45 สองครั้ง)
 * ถ้าไม่ตัด การนับย้อน "8 แท่ง = 2 ชั่วโมง" จะเหลือ 7 แท่งจริง
 * และการเทียบแท่งล่าสุดกับแท่งก่อนหน้าจะกลายเป็นเทียบแท่งเดียวกับตัวเอง
 * เก็บตัวหลังเพราะเป็นข้อมูลที่ใหม่กว่า
 */
export function dedupeBars(bars: Bar[]): Bar[] {
  const byTs = new Map<number, Bar>();
  for (const b of bars) byTs.set(b.ts, b);
  return [...byTs.values()].sort((a, b) => a.ts - b.ts);
}

export async function fetchGcf(interval = "15m", range = "60d"): Promise<Bar[]> {
  const data = await getJson<YahooChart>(`${GCF_URL}?interval=${interval}&range=${range}`, 40_000, TTL.history);
  const result = data.chart.result[0];
  const q = result.indicators.quote[0];

  const bars: Bar[] = [];
  result.timestamp.forEach((ts, i) => {
    const h = q.high[i], l = q.low[i], c = q.close[i], o = q.open[i];
    if (h === null || l === null || c === null) return;
    bars.push({ ts, o: o ?? c, h, l, c });
  });
  return dedupeBars(bars);
}

/**
 * ราคาสดกับราคาอ้างอิงเมื่อ 15–30 นาทีก่อน ในสเกลโบรกเกอร์ — ใช้ตัดสินทิศทางตอนนี้
 *
 * แท่งสุดท้ายคือแท่งที่ยังไม่ปิด (ราคาสด) จึงข้ามแท่งที่เพิ่งปิดไปใช้แท่งก่อนหน้านั้นอีกแท่ง
 * ถ้าใช้แท่งที่เพิ่งปิด ช่วงต้นแท่งใหม่ราคายังไม่ทันขยับ ผลต่างจะเป็นศูนย์ แล้วบอก "ทรงตัว"
 * ทั้งที่ราคากำลังร่วง — และ cron ยิงตรง :00 :15 :30 :45 พอดี จะเจอกรณีนี้แทบทุกรอบ
 *
 * ใช้ร่วมกันทั้งหน้าหลักและ cron จะได้ไม่คำนวณคนละแบบ
 */
export function liveQuote(p: PriceData): { price: number; ref: number } | undefined {
  if (p.bars.length < 3) return undefined;
  return { price: p.price, ref: p.bars[p.bars.length - 3].c };
}

export interface PriceData {
  /** ราคาปัจจุบันในสเกลโบรกเกอร์ */
  price: number;
  spot: number | null;
  basis: number;
  brokerOffset: number;
  /** แท่ง 15 นาที ปรับเป็นสเกลโบรกเกอร์แล้ว (ใช้หาแนวรับแนวต้าน + วาดกราฟ) */
  bars: Bar[];
  /** แท่ง 15 นาที 60 วันของ GC=F (ใช้ทำโปรไฟล์ความผันผวนรายชั่วโมง) */
  profileBars: Bar[];
  /** แท่งเทียนมาจากไหน — PAXG ใกล้ spot กว่า แต่ Binance บล็อก IP สหรัฐฯ */
  barSource: "PAXG" | "GC=F";
  /** ปรับฐานด้วยราคา spot จริงได้ไหม — ถ้าไม่ได้ ระดับราคาจะเชื่อถือไม่ได้ */
  calibrated: boolean;
  fetchedAt: number;
  errors: string[];
}

/**
 * โหลดราคา — ออกแบบให้ไม่ผูกกับแหล่งเดียว
 *
 * PAXG (Binance) ให้ค่าใกล้ spot ที่สุด (ห่าง ~$5) แต่ **Binance บล็อก IP สหรัฐฯ (HTTP 451)**
 * ซึ่ง Vercel รัน function ที่ภูมิภาค US เป็นค่าเริ่มต้น — เวอร์ชันแรกจึงใช้ได้แค่บนเครื่อง
 * ที่ IP ไทย พอขึ้นเซิร์ฟเวอร์กราฟหายทั้งหน้า
 *
 * จึงสำรองด้วย GC=F ซึ่งดึงมาอยู่แล้วสำหรับโปรไฟล์รายชั่วโมง (ไม่เปลืองรีเควสต์เพิ่ม)
 * ห่าง spot ~$35 แต่ปรับฐานด้วย spot จริงได้เหมือนกัน
 */
export async function loadPrice(brokerOffset = 0): Promise<PriceData | null> {
  const errors: string[] = [];

  const [spotRes, paxgRes, gcfRes] = await Promise.allSettled([
    fetchSpot(),
    fetchPaxg15m(),
    fetchGcf("15m", "60d"),
  ]);

  const spot = spotRes.status === "fulfilled" ? spotRes.value : null;
  if (spotRes.status === "rejected") errors.push(`ราคา spot: ${spotRes.reason}`);

  const gcfBars = gcfRes.status === "fulfilled" ? gcfRes.value : [];
  if (gcfRes.status === "rejected") errors.push(`GC=F: ${gcfRes.reason}`);

  const paxgBars = paxgRes.status === "fulfilled" ? paxgRes.value : [];
  if (paxgRes.status === "rejected") errors.push(`PAXG (Binance): ${paxgRes.reason}`);

  // เลือกแหล่งแท่งเทียน: PAXG ก่อนเพราะใกล้ spot กว่า ไม่ได้ค่อยใช้ GC=F
  const [rawBars, barSource]: [Bar[], PriceData["barSource"]] =
    paxgBars.length ? [paxgBars, "PAXG"] : [gcfBars, "GC=F"];

  if (!rawBars.length) {
    // ล้มทั้งสองแหล่ง — log ไว้ให้เห็นใน runtime log ของ Vercel ไม่งั้นจะ debug ไม่ได้เลย
    console.error("[price] ดึงแท่งเทียนไม่ได้ทั้ง PAXG และ GC=F:", errors.join(" | "));
    return null;
  }

  // ปรับแหล่งที่เลือก -> spot -> สเกลโบรกเกอร์
  const basis = spot ? spot - rawBars[rawBars.length - 1].c : 0;
  const shift = basis + brokerOffset;

  // PAXG ซื้อขาย 24/7 แท่งเสาร์อาทิตย์สภาพคล่องบางมากต้องตัดทิ้ง
  // (GC=F ไม่มีแท่งช่วงนั้นอยู่แล้ว กรองซ้ำก็ไม่เสียหาย)
  const bars = rawBars
    .filter((b) => isMarketOpen(b.ts))
    .map((b) => ({ ts: b.ts, o: b.o + shift, h: b.h + shift, l: b.l + shift, c: b.c + shift }));

  if (!bars.length) {
    console.error("[price] ไม่เหลือแท่งเทียนหลังกรองเวลาตลาด");
    return null;
  }

  if (errors.length) console.warn("[price] บางแหล่งใช้ไม่ได้:", errors.join(" | "));

  return {
    price: Math.round(bars[bars.length - 1].c * 100) / 100,
    spot: spot ? Math.round(spot * 100) / 100 : null,
    basis: Math.round(basis * 100) / 100,
    brokerOffset,
    bars,
    profileBars: gcfBars,
    barSource,
    calibrated: spot !== null,
    fetchedAt: Math.floor(Date.now() / 1000),
    errors,
  };
}

File: legacy-python/goldnews/price.py
"""ดึงราคาทอง 15 นาที และปรับให้เป็นสเกลราคาที่โบรกเกอร์แสดง

ทำไมต้องปรับฐาน:
    ทองล่วงหน้า COMEX (GC=F) แพงกว่า spot ราว $35 — ถ้าเอามาคำนวณแนวรับแนวต้านตรง ๆ
    เส้นจะเพี้ยนทั้งกระดานเมื่อวาดลงกราฟ XAUUSD ของโบรกเกอร์

จึงใช้:
    PAXG (โทเคนทองคำ 1 ออนซ์ ซื้อขาย 24 ชม.) เป็นแท่งเทียน 15 นาที  ห่าง spot ~$5
    + ปรับฐานด้วยราคา spot จริง  -> เหลือคลาดเคลื่อน ~$1-2
    + broker_offset จาก config  -> ตรงกับกราฟที่ผู้ใช้ดูจริง
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from datetime import datetime, timedelta
from pathlib import Path

from .fetch import CACHE_FILE, TH_TZ, UTC, _UA, _http_get

SPOT_URL = "https://api.gold-api.com/price/XAU"
PAXG_URL = "https://api.binance.com/api/v3/klines?symbol=PAXGUSDT&interval=15m&limit={limit}"
GCF_URL = "https://query1.finance.yahoo.com/v8/finance/chart/GC=F?interval={interval}&range={range}"

PRICE_CACHE = CACHE_FILE.parent / "price.json"

# ตลาดทองปิดช่วงสุดสัปดาห์ แต่ PAXG ซื้อขายตลอด — แท่งเสาร์อาทิตย์สภาพคล่องบางมาก ต้องตัดทิ้ง
# (ปิดเสาร์ ~04:00 น. ไทย เปิดจันทร์ ~05:00 น. ไทย)
def _is_market_open(dt: datetime) -> bool:
    weekday = dt.weekday()  # 0=จันทร์ 5=เสาร์ 6=อาทิตย์
    if weekday == 5:
        return dt.hour < 4
    if weekday == 6:
        return False
    if weekday == 0:
        return dt.hour >= 5
    return True


def fetch_spot() -> float:
    """ราคา spot ทองคำล่าสุด (ดอลลาร์ต่อออนซ์)"""
    return float(json.loads(_http_get(SPOT_URL))["price"])


def fetch_paxg_15m(limit: int = 1000) -> list[dict]:
    """แท่งเทียน 15 นาทีของ PAXG — ยังไม่ปรับฐาน"""
    rows = json.loads(_http_get(PAXG_URL.format(limit=limit)))
    return [
        {
            "ts": int(row[0]) // 1000,
            "o": float(row[1]),
            "h": float(row[2]),
            "l": float(row[3]),
            "c": float(row[4]),
        }
        for row in rows
    ]


def fetch_gcf(interval: str = "15m", range_: str = "60d") -> list[dict]:
    """แท่งเทียนทองล่วงหน้า COMEX (GC=F) เรียงตามเวลา

    ใช้ GC=F เพราะเป็นตลาดทองจริงที่มีเวลาเปิดปิดจริงและข่าววิ่งเข้าโดยตรง
    ราคาห่าง spot ราว $35 แต่ไม่มีผล เพราะงานที่ใช้วัดแค่ "ส่วนต่าง" กับ "ช่วงกว้าง"
    ไม่ได้ใช้ระดับราคา

    ขอบเขตที่ Yahoo ให้ (ทดสอบแล้ว): 1m=7วัน · 5m=60วัน · 15m=60วัน · 1h=2ปี
    """
    payload = json.loads(_http_get(GCF_URL.format(interval=interval, range=range_)))
    result = payload["chart"]["result"][0]
    quote = result["indicators"]["quote"][0]

    bars = []
    for i, ts in enumerate(result["timestamp"]):
        high, low, close = quote["high"][i], quote["low"][i], quote["close"][i]
        if high is None or low is None or close is None:
            continue
        bars.append({"ts": int(ts), "o": quote["open"][i], "h": high, "l": low, "c": close})
    bars.sort(key=lambda b: b["ts"])
    return bars


def fetch_gcf_15m(range_: str = "60d") -> list[dict]:
    """แท่ง 15 นาที — ใช้ทำโปรไฟล์ความผันผวนรายชั่วโมง"""
    return fetch_gcf("15m", range_)


def fetch_gcf_5m(range_: str = "60d") -> list[dict]:
    """แท่ง 5 นาที — ละเอียดพอวัดปฏิกิริยาข่าวที่ +5/+10/+15 นาที (Reaction Lab)"""
    return fetch_gcf("5m", range_)


def fetch_gcf_1h(range_: str = "2y") -> list[dict]:
    """แท่งรายชั่วโมง 2 ปี — ใช้ backfill ขอบเขต 60 นาทีย้อนหลังไกล"""
    return fetch_gcf("1h", range_)


def _read_cache() -> dict | None:
    if not PRICE_CACHE.exists():
        return None
    try:
        return json.loads(PRICE_CACHE.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return None


def load(broker_offset: float = 0.0) -> dict:
    """คืน {price, bars, profile_bars, basis, spot, stale, errors}

    bars         = แท่ง 15 นาที ปรับเป็นสเกลราคาโบรกเกอร์แล้ว (ใช้หาแนวรับแนวต้าน)
    profile_bars = แท่ง 15 นาที 60 วันของ GC=F (ใช้ทำโปรไฟล์ความผันผวนรายชั่วโมง)
    price        = ราคาปัจจุบันในสเกลโบรกเกอร์
    """
    errors: list[str] = []
    spot = None
    try:
        spot = fetch_spot()
    except (urllib.error.URLError, OSError, ValueError, KeyError) as exc:
        errors.append(f"ราคา spot: {exc}")

    raw_bars: list[dict] = []
    try:
        raw_bars = fetch_paxg_15m()
    except (urllib.error.URLError, OSError, ValueError, KeyError) as exc:
        errors.append(f"PAXG 15 นาที: {exc}")

    profile_bars: list[dict] = []
    try:
        profile_bars = fetch_gcf_15m()
    except (urllib.error.URLError, OSError, ValueError, KeyError, IndexError) as exc:
        errors.append(f"GC=F 60 วัน: {exc}")

    if not raw_bars:
        cached = _read_cache()
        if not cached:
            raise RuntimeError("ดึงราคาไม่สำเร็จ และไม่มีไฟล์ cache:\n  - " + "\n  - ".join(errors))
        errors.append("ใช้ราคาสำรองจากไฟล์ cache")
        cached["stale"] = True
        cached["errors"] = errors
        return cached

    # ปรับ PAXG -> spot -> สเกลโบรกเกอร์
    basis = (spot - raw_bars[-1]["c"]) if spot else 0.0
    shift = basis + broker_offset

    bars = [
        {
            "ts": b["ts"],
            "o": b["o"] + shift,
            "h": b["h"] + shift,
            "l": b["l"] + shift,
            "c": b["c"] + shift,
        }
        for b in raw_bars
        if _is_market_open(datetime.fromtimestamp(b["ts"], TH_TZ))
    ]

    data = {
        "price": round(bars[-1]["c"], 2),
        "spot": round(spot, 2) if spot else None,
        "basis": round(basis, 2),
        "broker_offset": broker_offset,
        "bars": bars,
        "profile_bars": profile_bars,
        "fetched_at": int(datetime.now(UTC).timestamp()),
        "stale": False,
        "errors": errors,
    }

    PRICE_CACHE.parent.mkdir(parents=True, exist_ok=True)
    PRICE_CACHE.write_text(json.dumps(data), encoding="utf-8")
    return data
